pkg/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,290 @@
+			tableRemovedCount, symbolRemovedCount,
+			len(s.symbolTables), s.symbolCountTotal)
+	}
+}
+
+// genKey generates a unique key for an executable given inode and mtime.
+// This key is used to cache symbol tables.
+func genKey(ino uint64, mtimeSec int64, mtimeNsec uint32) string {
+	return fmt.Sprintf("ino=%d mtime=%d.%d", ino, mtimeSec, mtimeNsec)
+}
+
+// lookupByAddr returns the symbol name for the given address.
+func (e *symbolTable) lookupByAddr(address uint64) string {
+	// Similar to a trivial binary search, but each symbol is a range.
+	n, found := slices.BinarySearchFunc(e.symbols, address, func(a *symbol, b uint64) int {
+		if a.value <= b && a.value+a.size > b {
+			return 0
+		}
+		if a.value > b {
+			return 1
+		}
+		if a.value < b {
+			return -1
+		}
+		return 0
+	})
+	if found {
+		return e.symbols[n].name
+	}
+	return "[unknown]"
+}
+
+type Task struct {
+	// Pid of the process that we want to resolve symbols for.
+	Pid uint32
+
+	// Name of the task for logging purposes. Optional.
+	Name string
+
+	// ContainerPid is the pid 1 of the container that the process is in. Optional.
+	ContainerPid uint32
+
+	// Properties of the executable to check if the symbol table is still valid.
+	Ino       uint64
+	MtimeSec  int64
+	MtimeNsec uint32
+}
+
+func (s *Symbolizer) Resolve(task Task, addresses []uint64) ([]string, error) {
+	res := make([]string, len(addresses))
+
+	if len(addresses) == 0 {
+		return res, nil
+	}
+
+	exeKey := genKey(task.Ino, task.MtimeSec, task.MtimeNsec)
+	s.lock.RLock()
+	table, ok := s.symbolTables[exeKey]
+	if ok {
+		table.timestamp = time.Now()
+		for idx, addr := range addresses {
+			res[idx] = table.lookupByAddr(addr)
+		}
+		s.lock.RUnlock()
+		return res, nil
+	}
+	s.lock.RUnlock()
+
+	var err error
+	table, err = newSymbolTable(task.Pid, exeKey)
+	if err != nil {
+		return nil, fmt.Errorf("creating new symbolTable for %q: %w", task.Name, err)
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if len(table.symbols)+s.symbolCountTotal > maxSymbolCountTotal {
+		return nil, fmt.Errorf("too many symbols in all symbol tables: %d (max: %d)",
+			len(table.symbols)+s.symbolCountTotal, maxSymbolCountTotal)
+	}
+
+	// Most gadgets won't use the symbolizer, so don't start the prune loop until we need it.
+	if !s.pruneLoopStarted {
+		s.pruneLoopStarted = true
+		go s.pruneLoop()
+	}
+
+	s.symbolTables[exeKey] = table
+	s.symbolCountTotal += len(table.symbols)
+
+	log.Debugf("symbol table for %q (pid %d) loaded: %d symbols (total: %d symbol tables with %d symbols)",
+		task.Name, task.Pid, len(table.symbols), len(s.symbolTables), s.symbolCountTotal)
+
+	for idx, addr := range addresses {
+		res[idx] = table.lookupByAddr(addr)
+	}
+	return res, nil
+}
+
+func newSymbolTable(pid uint32, expectedExeKey string) (*symbolTable, error) {
+	var symbols []*symbol
+
+	path := fmt.Sprintf("%s/%d/exe", host.HostProcFs, pid)
```

Comment:
This does not work on minikube with the docker driver:
- the `pid` comes from ebpf, so the pid number is relative to the top-level pid namespace
- HostProcFs (`/host/proc`) is a procfs mount for the minikube host pid namespace

The CI fails because of that:
![image](https://github.com/user-attachments/assets/de45a42e-8c1d-4576-a778-036b48c90e33)